### PR TITLE
Feature: Diagram type

### LINKED ISSUE #401: Bug when drag and drop component from a diagram in root
**Describe the bug**

When we create a diagram at the root, and drag and drop a component inside Draw view, a new file `null` is created inside Text view, instead of using the file related to that diagram. This bug seems to have popped since this bug fix https://github.com/ditrit/leto-modelizer/issues/377 (PR https://github.com/ditrit/leto-modelizer/pull/394) and might be link to a wrong filePath.

**To Reproduce**

demo

https://github.com/ditrit/leto-modelizer/assets/62500081/d155e35a-d45b-4f3d-9261-4571cd6adae5

**Expected behavior**

Same behavior as when we click on a diagram instead of drag and drop (component's code must be written inside the corresponding file)

**Application:**

- leto version 1.3.0

**Additional context**

Checkout `modelPath` parameter inside`renderModel` function inside `PluginManager` composable

### PR DESCRIPTION
**Pull request by justinehell**
Resolve #401 and #415

### CODE REVIEW

File: src/components/form/CreateModelForm.vue [hunk 1]
```diff
@@ -99,12 +102,35 @@ const props = defineProps({
...
+
+/**
+ * Check if new diagram to create has a valid path.
+ * @returns {boolean | string} Return true if the value is a valid diagram path,
+ * otherwise the translated error message.
+ */
+function isValidDiagramPath() {
+  return getPluginByName(modelPlugin.value)
+    .isParsable(new FileInformation({ path: modelLocation.value }))
+    ? null : t('errors.models.notParsable');
```
Comment (k-y0u) [inline src/components/form/CreateModelForm.vue:132]: should return a bool according to the description.

File: src/components/form/CreateModelForm.vue [hunk 2]
```diff
@@ -160,16 +180,23 @@ async function onSubmit() {
...
 
 onMounted(async () => {
   getAllModels(props.projectName).then((array) => {
     models.value = array;
   });
-  allFolders.value = await getProjectFolders(props.projectName);
 });
 </script>
+
+<style lang="scss">
```
Comment (k-y0u) [inline src/components/form/CreateModelForm.vue:198]: why is needed now but not before ?

File: tests/unit/components/form/CreateModelForm.spec.js [hunk 3]
```diff
@@ -64,19 +68,121 @@ describe('Test component: CreateModelForm', () => {
...
-      wrapper.vm.modelName = null;
+    it('should set modelPath to empty string if plugin.isFolderTypeDiagram is true', () => {
+      wrapper.vm.modelPath = null;
+
+      wrapper.vm.onPluginChange();
+
+      expect(wrapper.vm.modelPath).toEqual('');
+    });
+
+    it('should set modelPath equals to defaultFileName if plugin.isFolderTypeDiagram is false', () => {
```
Comment (k-y0u) [inline tests/unit/components/form/CreateModelForm.spec.js:146]: equal

File: changelog.md [hunk 4]
```diff
@@ -17,12 +17,17 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+  * Work with [plugin-core](https://github.com/ditrit/leto-modelizer-plugin-core/tree/0.22.0) version 0.22.0.
+  * Work with [terrator-plugin](https://github.com/ditrit/terrator-plugin/tree/0.7.1) version 0.7.1.
+  * Work with [githubator-plugin](https://github.com/ditrit/githubator-plugin/tree/0.3.0) version 0.3.0.
 
 ### Fixed
 
 * Fix bug on click on component definition from plugins with single file diagrams, see [this issue](https://github.com/ditrit/leto-modelizer/issues/407).
 * Fix bug about rewritten file on checkout, see [this issue](https://github.com/ditrit/leto-modelizer/issues/422).
 * Fix bug about required management for RegExp rule, see [this issue](https://github.com/ditrit/leto-modelizer/issues/387).
+* Fix bug on different behavior of the component between a click and a drag and drop, see [this issue](https://github.com/ditrit/leto-modelizer/issues/401) and [this issue](https://github.com/ditrit/leto-modelizer/issues/415).
```
Comment (k-y0u) [inline changelog.md:30]: Fix bugs about component creation on a click or a drag&drop

**Review by k-y0u — CHANGES_REQUESTED** [review]

File: changelog.md (hunk 4 above)
Comment (justinehell) [reply changelog.md:30]: done

File: tests/unit/components/form/CreateModelForm.spec.js (hunk 3 above)
Comment (justinehell) [reply tests/unit/components/form/CreateModelForm.spec.js:146]: done

File: src/components/form/CreateModelForm.vue (hunk 1 above)
Comment (justinehell) [reply src/components/form/CreateModelForm.vue:132]: done

File: src/composables/PluginManager.js [hunk 5]
```diff
@@ -186,20 +186,37 @@ export async function renderModel(projectId, modelPath, plugin) {
...
 
-      return deleteProjectFile(projectId, file.path);
-    }),
+  if (renderFiles.length === 0) {
+    renderFiles.push(filesToDelete.pop());
+  }
+
+  renderFiles.push(...allUnparsableFiles);
+
+  return Promise.allSettled(
```
[the remark this answers is not in the document]
Comment (justinehell) [reply src/composables/PluginManager.js:216]: done

File: src/composables/PluginManager.js [hunk 6]
```diff
@@ -186,20 +186,37 @@ export async function renderModel(projectId, modelPath, plugin) {
...
     files.filter((file) => plugin.isParsable(file)),
   );
 
-  return Promise.allSettled(
-    renderFiles.map((file) => {
-      if (file.content) {
-        return writeProjectFile(projectId, file);
+  const renderFiles = [];
+  const filesToDelete = [];
+  const allUnparsableFiles = [];
```
[the remark this answers is not in the document]
Comment (justinehell) [reply src/composables/PluginManager.js:196]: done

File: src/composables/PluginManager.js [hunk 7]
```diff
@@ -186,20 +186,37 @@ export async function renderModel(projectId, modelPath, plugin) {
...
     diagramFile,
     config,
     files.filter((file) => plugin.isParsable(file)),
   );
 
-  return Promise.allSettled(
-    renderFiles.map((file) => {
-      if (file.content) {
-        return writeProjectFile(projectId, file);
+  const renderFiles = [];
```
[the remark this answers is not in the document]
Comment (justinehell) [reply src/composables/PluginManager.js:194]: done

File: src/composables/PluginManager.js [hunk 8]
```diff
@@ -186,20 +186,37 @@ export async function renderModel(projectId, modelPath, plugin) {
     diagramFile.path = '';
   }
 
-  const renderFiles = plugin.render(
+  const allFiles = plugin.render(
```
[the remark this answers is not in the document]
Comment (justinehell) [reply src/composables/PluginManager.js:188]: done

File: tests/unit/components/card/ComponentDefinitionCard.spec.js [hunk 9]
```diff
@@ -139,25 +127,10 @@ describe('Test component: ComponentDefinitionCard', () => {
...
-      addNewComponent.mockClear();
-
-      getModelFiles.mockImplementation(() => Promise.resolve([{ path: 'test' }]));
-
+    it('should call plugin.addComponent when isTemplate is false', async () => {
       await wrapper.vm.onClickItem();
 
-      expect(addNewComponent).toHaveBeenCalledTimes(1);
-      expect(addNewComponent.mock.calls[0][2]).toEqual('test');
+      expect(wrapper.vm.plugin.addComponent).toHaveBeenCalled();
```
[the remark this answers is not in the document]
Comment (justinehell) [reply tests/unit/components/card/ComponentDefinitionCard.spec.js:133]: done

File: tests/unit/components/card/ComponentDefinitionCard.spec.js [hunk 10]
```diff
@@ -169,12 +142,20 @@ describe('Test component: ComponentDefinitionCard', () => {
...
 
       await wrapper.vm.onClickItem();
 
       expect(addNewTemplateComponent).toHaveBeenCalledTimes(1);
     });
+
+    it('should call plugin.draw and renderModel', async () => {
+      await wrapper.vm.onClickItem();
+
+      expect(wrapper.vm.plugin.draw).toBeCalledWith('root');
```
[the remark this answers is not in the document]
Comment (justinehell) [reply tests/unit/components/card/ComponentDefinitionCard.spec.js:156]: done

File: src/components/form/CreateModelForm.vue (hunk 2 above)
Comment (justinehell) [reply src/components/form/CreateModelForm.vue:198]: it's not needed, but better for UI
(removed `lang="scss"` and add `scoped`)

File: src/composables/PluginManager.js [hunk 11]
```diff
@@ -186,20 +186,37 @@ export async function renderModel(projectId, modelPath, plugin) {
...
       }
+    } else {
+      allUnparsableFiles.push(file);
+    }
+  });
 
-      return deleteProjectFile(projectId, file.path);
-    }),
+  if (renderFiles.length === 0) {
+    renderFiles.push(filesToDelete.pop());
```
[the remark this answers is not in the document]
Comment (justinehell) [reply src/composables/PluginManager.js:211]: We have to keep at least one file so that the diagram is still displayed in the diagramList

**Review by Zorin95670 — APPROVED** [review]

File: src/composables/PluginManager.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (k-y0u) [reply src/composables/PluginManager.js]: Should it be the deafaultFileName file, empty ? Should it be render by the plugin ?

File: src/components/form/CreateModelForm.vue (hunk 2 above)
Comment (k-y0u) [reply src/components/form/CreateModelForm.vue]: ok but better keep `lang="scss"`

Comment (justinehell) [reply src/components/form/CreateModelForm.vue]: ok

File: src/composables/PluginManager.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (justinehell) [reply src/composables/PluginManager.js]: updated

**Review by Zorin95670 — APPROVED** [review]

File: src/composables/PluginManager.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (k-y0u) [reply src/composables/PluginManager.js]: not satisfying

**Pull request merged**